### ISSUE TITLE
Langevin thermostat does not reproduce stokes einstein

### ISSUE THREAD

**Issue by rbardak** [espressomd/espresso#938]
This is related to #904 , I decided to open a new issue since this test did not cover correlator checkpointing, nor other features used in the simulations there. I used the recent 81d5c9a6bc49e48902880691af488c85be475d02 I ran the script https://github.com/rbardak/uploads/blob/master/stokes-einstein.py to measure the diffusion constant in a minimal system. I used different combinations of kT and gamma in an attempt to reproduce the Stokes-Einstein relation. 
![stokes-einstein-comparison](https://cloud.githubusercontent.com/assets/11994040/20489453/f7c5af04-b00a-11e6-8bc7-22a5235d6e56.png)
As you can see in the above figure, the measurements are systematically wrong. Earlier I reported these measurements were off by a factor of sqrt(3), however I was mistaken, they are actually off by a factor of sqrt(3)/2.

I ran a similar script with a sample input with the CPU LB of the same branch/hash (no sweep, since it took considerably longer). This produced the proper diffusion constant by three order of magnitude accuracy. This rules out bugs in the correlator, which confirms that the langevin thermostat is the cause of the problem.

All simulations ran on a single cpu core to rule out secondary influences to the active parts of the code.

P.S. If you try to reproduce my findings on the CPU LB you might run into an issue with the ghost communicator, I used an unorthodox workaround to get the LB running properly.

**Comment by rbardak** [comment]
CPU LB results for reference:
![stokes-einstein-comparison-lb](https://cloud.githubusercontent.com/assets/11994040/20493738/4cbb4adc-b01a-11e6-8ae7-fc6565408930.png)

here, the deviation is less than 5%. This seems acceptable, especially considering that one may expect errors from the LB with high diffusion coefficients.
The corresponding script: https://github.com/rbardak/uploads/blob/master/stokes-einstein-lb.py
It's the same as before, except for two additions and one change.

**Comment by jonaslandsgesell** [comment]
Could you maybe check wether Espresso correctly worked before we switched to the boost RNG? This is where i would start.

**Comment by jdegraaf** [comment]
Dear All,

I concur, the current master does not give the right result (time on x-axis, MSD on y-axis):

![analysis_langevin_msd](https://cloud.githubusercontent.com/assets/2853791/20528189/1344b29e-b0c3-11e6-800e-79f47668a78c.png)

Blue dots are the measured data (mean of 200 runs), black curves are trends t^2 and t, magenta is the MSD for the expected diffusion coefficient, cyan the measured one. This is indeed well approximated by a 2/sqrt(3) increase. LB should not reproduce the expected diffusion coefficient exactly, due to particle-fluid coupling, so 5% error is quite acceptable, indeed. 

The result is consistently wrong over the MSD and VACF (velocity autocorrelation function).

Robin, can you check if this error has crept in some time between my fix of the Langevin thermostat to make it take a value for the friction coefficient, rather then the relaxation time, which was around:

20839f97a5fe4a9684725020a09a9f95f7a616d6

I would be highly surprised if that was what broke the behavior, as I thoroughly checked the implementation at the time. Also, I believe that I implemented a different translational diffusion coefficient along the three principal axes of the particle, but this does not appear to be anywhere in the current master (it is there for the rotation, but this was implemented by someone else; strange...).

I presume the bug was introduced before 

da8627f4fbbc0b1ca35b87ae0bc932ded3b8f7f0

Since that one says "Increased tolernace of langevin_per_particle by 15%" because of constant errors being thrown. However, it could be that that 15% accounts for the factor 2/sqrt(3). 

Best Wishes,

Joost

**Comment by jdegraaf** [comment]
Dear All,

I performed more checks, this time with the Python master

9180580414be31e0ac85b0bdb73749f998762c48

I found that here the Langevin thermostat performed as expected, see the two scripts with which I generated the data

[drag.txt](https://github.com/espressomd/espresso/files/619326/drag.txt)
[langevin.txt](https://github.com/espressomd/espresso/files/619327/langevin.txt)
[myconfig.txt](https://github.com/espressomd/espresso/files/619347/myconfig.txt)

The extension was changed to allow for upload to github. The first file can be used to show that dissipation works without fluctuations. The second can be used to show that fluctuation-dissipation works. So, my guess is that the problem lies in the TcL interface somehow, or that the error is in some other way repaired in the Python interface. The error is at most 1% from the expected value from the VACF and AVACF (which are more sensible to use than the MSD)

![translational_diffusion_from_vacf](https://cloud.githubusercontent.com/assets/2853791/20715306/a97d022e-b646-11e6-896e-c6749ab44109.png)

Above, the magenta line is the fit, the black one the expected value for the translational diffusion coefficient.

![rotational_diffusion_from_avacf](https://cloud.githubusercontent.com/assets/2853791/20715299/a6333d90-b646-11e6-97ab-bac4a058f2a5.png)

Above, the magenta line is the fit, the black one the expected value for rotational diffusion coefficient.

![msd](https://cloud.githubusercontent.com/assets/2853791/20715608/a11011a2-b647-11e6-94b4-6aaab74298a0.png)

Above, the cyan line is the fit to the translational diffusion coefficient. The deviation from the expected value is 3% (averaged over 25 independent runs). The black lines are guides to the eye for the power-law decay and transition from one scaling behaviour to the other.

Perhaps adding another feature to the set in the "myconfig.txt(hpp)" file will break things, but for now the Python branch seems to do the job just fine.

Best Wishes,

Joost

**Comment by rbardak** [comment]
I ran a few more tests and as far as I can tell I have isolated the problem to the ROTATION feature. If the feature is not activated, the error occurs. Looking at your config, you used the ROTATION feature, therefore you got the right results. 
I will start working on a fix tomorrow, as this is most likely an error somewhere in the ifdef jungle of the thermostat code.

**Comment by jdegraaf** [comment]
Hi Robert,

Perhaps, but the previous test I carried out was with ROTATION included as
well, only using the TcL interface. I therefore suspect that it is more
likely a TcL thing. However, you could try to run the scripts I attached
without ROTATION and see if they give wrong results << you'll need to strip
out the rotation related bits, of course >>. I would definitely do that
before hacking in the code, because I'm not sure if having that feature or
not is the issue.

Best Wishes,

Joost

**Comment by rbardak** [comment]
> I therefore suspect that it is more likely a TcL thing.

All of my tests are carried out with the python interface. I find it unlikely that erroneous code is present in both interfaces, which on top of that depends on the myconfig. Therefore the simplest conclusion would be that the error is present in the core, most likely in a place where a single misplaced line wrecks the thermostat.

Regarding your previous results using TCL, could you provide the script and config in this case? If what you say about your config is true then the error might also depend on what features are used within the script.

**Comment by jdegraaf** [comment]
Hello Robin,

Please find attached the relevant files, however, this script
(swimmer_measure.lvne.tcl) is considerably more complex than the one I used
for the Langevin tests with Python, as it has checkpointing and a  bunch of
other stuff to simulate active particles. But as you can see, my
myconfig.hpp file did have rotation put in at the time of the evaluation. I
used the data sets for which I did not have activity switched on to produce
the MSD curve I showed on github.

Best Wishes,

Joost

**Comment by RudolfWeeber** [comment]
I just checked with a debugger that the relevant prefactors in thermostat.hpp:308
(langevin_pref1_temp and langevin_pref2_temp)
are the same (-100, 84.something) for
* default myconfig
* default myconfig with rotation added
* default myconfig with rotational_inertia added

Robin, with which configs exactly did you check?

**Comment by rbardak** [comment]
I checked with a completely empty config and a config defining ROTATION
only.
For the lb measurements I only defined LB.

**Comment by RudolfWeeber** [comment]
On Thu, Dec 01, 2016 at 07:25:01AM -0800, Robin Bardakcioglu wrote:
> I checked with a completely empty config and a config defining ROTATION
> only.
> For the lb measurements I only defined LB.
The prefactors are the same for an empty myconfig.
Are you sure nothing but the content of myconfig changed between the two experiments. It's a bit unlikely that a different diffusion coeff comes out, if the prefactors for friction and noise terms are the same.

**Comment by rbardak** [comment]
Yes I thoroughly checked that nothing changed except the config. I ran the
script that I supplied in my first post for both builds and got different
results.

**Comment by jdegraaf** [comment]
Dear Robin and Rudolf,

I'm not quite following the discussion, are you saying that there is an
issue with the langevin parameters getting random junk written into them
depending on the specific compiler flags set?

Best Wishes,

Joost

**Comment by rbardak** [comment]
We are discussing my observation that the occurance of the problem depends
on the myconfig. Rudolf is currently checking if the myconfig influences
the values of the langevin prefactors by virtue of misplaced ifdefs or
unexpected behaviour of the code.

**Comment by RudolfWeeber** [comment]
So, now I also compared the prefactors in thermostat.hpp for an
empty myconfig, and also for kT !=1. They are identical, whether or
not rotation is compiled in.

**Comment by RudolfWeeber** [comment]
I cannot reproduce the wrong diffusion coefficient.
The following script checks both the msd via Stokes-Einstein and the vacf via Grren Kubo. 
Both are correct.
Robin, can you please run your script longer by a factor of 50 (you can increase the time step). With a simulation time of 40 000 as in your script, I don't get good enough statistics. The MSD doesn't look noisy, but has the wrong shape.

I ran this for default myconfig and default myconfig+rotation with similar results


gamma=2.4
kT=1.37
dt=0.05

import espressomd
from espressomd.correlators import *
from espressomd.observables import *
from scipy import *

s=espressomd.System()
p=s.part.add(pos=(0,0,0),id=0)
s.time_step=dt
s.thermostat.set_langevin(kT=kT,gamma=gamma)
s.cell_system.skin =0.4
s.integrator.run(1000)

pos_obs=ParticlePositions(ids=(0,))
vel_obs=ParticleVelocities(ids=(0,))



c_pos = Correlator(obs1=pos_obs, tau_lin = 16, tau_max = 1000., dt = 10*dt,
        corr_operation="square_distance_componentwise", compress1="discard1")
c_vel = Correlator(obs1=vel_obs, tau_lin = 16, tau_max = 20., dt = dt,
        corr_operation="scalar_product", compress1="discard1")
s.auto_update_correlators.add(c_pos)
s.auto_update_correlators.add(c_vel)
s.integrator.run(10000000)

c_pos.finalize()
c_vel.finalize()

savetxt("msd.dat",c_pos.result())
savetxt("vacf.dat",c_vel.result())

# Integral of vacf via Green-Kubo
#D= 1/3 int_0^infty <v(t_0)v(t_0+t)> dt

vacf=c_vel.result()
#Integrate w. trapez rule
I=trapz(vacf[:,2],vacf[:,0])
print "Ratio of measured and expected diffusion coeff from Green-Kubo:", 1./3.*I/(kT/gamma)

# Check msd
msd=c_pos.result()

expected_msd = lambda x: 2.*kT/gamma *x

print "Ratio of expected and measured msd"
print "#time ratio_x ratio_y ratio_z"
for i in range(16,msd.shape[0],16):
  print msd[i,0],msd[i,2:5]/expected_msd(msd[i,0])

**Comment by rbardak** [comment]
I ran the script as per @RudolfWeeber 's suggestion and I seem to produce the correct diffusion constant, both with MSD and VACF. To the best of my knowledge, my original tests ran long enough to produce good statistics and the MSD was properly converged, however it seems I was mistaken. 

As to the correlation with the ROTATION feature, I have gotten to the bottom of this. In light that the statistics weren't good enough for the MSD, running the simulation with the RNG being in a different state will influence the results. The ROTATION feature has the effect that for every integration step three additional random numbers are drawn for the thermalisation of the rotational degrees of freedom. I emulated this effect by discarding three random numbers after every integration step and compiled espresso without ROTATION. This produced the same results as the ones I have gotten with the feature activated. This explains the random correlation to the rotation feature.

I suppose there are three open questions left:
1) @jdegraaf 's reproduction of the error via tcl and the observation of the same deviation factor consistent over both MSD and VACF.
2) Why are these errors consistent over different sets of kT and gamma? I would expect this to be more of a random nature when statistics are bad.
3) why did the LB produce reliable results with the same simulation duration?

**Comment by jdegraaf** [comment]
Hi Robin,

That's interesting. For my MSD curves (the ones I ran with the TcL version
of ESPResSo) I can assure you they were properly converged, both in the
MSD, VACF, and AVACF. I also used a different seed for the random number
generator every time. However, as I mentioned before, I have not managed to
reproduce the error in the Python version, where I get the right result. LB
works very different so I would not expect it to be able to compare LV and
LB.

Best,

Joost

**Comment by RudolfWeeber** [comment]
I now also wrote a tcl version of the python msd and vacf checker posted in a previous comment.
For the default myconfig and current (tcl) master, I get the correct beahaviour.
Joost, can you please check the script, and then run it with your build and report the results?
On my machine, this runs a few minutes.

Tcl script:
---------------------------------------------------------
set gamma 2.4
set kT 1.37
set dt 0.05

part 0 pos 0 0 0
setmd time_step $dt
thermostat langevin $kT $gamma
setmd skin 0.4
integrate 1000

set pos_obs [observable new particle_positions all]
set vel_obs [observable new particle_velocities all]



set c_pos [correlation new obs1 $pos_obs tau_lin 16 tau_max 1000 dt [expr 10*$dt] corr_operation square_distance_componentwise compress1 discard1]
correlation $c_pos autoupdate start
set c_vel [correlation new obs1 $vel_obs tau_lin 16 tau_max 20 dt $dt corr_operation scalar_product compress1 discard1]
correlation $c_vel autoupdate start

integrate 10000000

correlation $c_pos finalize
correlation $c_vel finalize

set f [open msd_tcl.dat w]
puts $f [correlation $c_pos print]
close $f

set f [open vacf_tcl.dat w]
puts $f [correlation $c_vel print]
close $f
-----------------------------------------
Post processing the output (bash script)
---------------------------------------------

for f in msd_tcl.dat vacf_tcl.dat
do
 sed -i -e 's/ { //g' $f
 sed -i -e 's/ }//g' $f
 sed -i -e 's/  / /g' $f
done
-------------------------------------------------------------
Analysis (python). Here values close to unity should be reported
---------------------------------------------------------------
from scipy import *

gamma=2.4
kT=1.37


msd=genfromtxt("msd_tcl.dat")
vacf=genfromtxt("vacf_tcl.dat")
# Integral of vacf via Green-Kubo
#D= 1/3 int_0^infty <v(t_0)v(t_0+t)> dt

#Integrate w. trapez rule
I=trapz(vacf[:,2],vacf[:,0])
print "Ratio of measured and expected diffusion coeff from Green-Kubo:", 1./3.*I/(kT/gamma)

# Check msd

expected_msd = lambda x: 2.*kT/gamma *x

print "Ratio of expected and measured msd"
print "#time ratio_x ratio_y ratio_z"
for i in range(16,msd.shape[0],16):
  print msd[i,0],msd[i,2:5]/expected_msd(msd[i,0])

**Comment by jdegraaf** [comment]
Hi Rudolf,

I check this with the version of ESPResSo I was running the tests with and
it did not seem to fail, i.e., they predicted the correct value via
Green-Kubo to within 3%. I ran these with several different feature
combinations. I guess this implies that any problem was due to the
checkpointing. That makes some sense, since I did carefully check the
Langevin thermostat when I made the changes to it more than a year ago.
Still pretty peculiar that this resulted in a 2/sqrt(3) deviation, which
seem to me to be a bit too 'nice' a number to be a coincidence. I have some
longer, far more complicated simulations running with the latest Master,
which I will analyze tomorrow and I'll let you know what comes out of that.

Best Wishes,

Joost
